### PR TITLE
Glukkon::CreateFromSaveState_442830() && Glukkon::vGetSaveState_444B90()

### PR DESCRIPTION
**Pull request by Varrok**
💾 ◀️ 💻 

💾 ▶️ 💻

### CODE REVIEW

File: Source/AliveLib/Glukkon.cpp [hunk 1]
```diff
@@ -119,6 +129,130 @@ TintEntry stru_5546B4[18] =
...
+signed int CC Glukkon::CreateFromSaveState_442830(const BYTE* pData)
+{
+    const Glukkon_SaveState* pSaveState = reinterpret_cast< const Glukkon_SaveState* >( pData );
+    auto pTlv = static_cast< Path_Glukkon* >( sPath_dword_BB47C0->TLV_From_Offset_Lvl_Cam_4DB770(pSaveState->field_44_tlvInfo) );
+
+    switch (gMap_5C3030.sCurrentLevelId_5C3030)
+    {
+        case LevelIds::eFeeCoDepot_5:
+        case LevelIds::eFeeCoDepot_Ender_12:
+            if (!ResourceManager::GetLoadedResource_49C2A0(1835626049, 802, 0, 0))
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.cpp]: these can be replaced with constants that exist in the resource manager

File: Source/AliveLib/Glukkon.cpp [hunk 2]
```diff
@@ -119,6 +129,130 @@ TintEntry stru_5546B4[18] =
...
+    }
+
+    auto pGlukkon = alive_new<Glukkon>();
+    if (pGlukkon)
+    {
+        pGlukkon->ctor_43F030(pTlv, pSaveState->field_44_tlvInfo);
+    }
+    else
+    {
+        pGlukkon = 0;
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.cpp]: this is redundant, if !pGlukkon means it must be 0/nullptr already

File: Source/AliveLib/Glukkon.cpp [hunk 3]
```diff
@@ -119,6 +129,130 @@ TintEntry stru_5546B4[18] =
...
+    pGlukkon->field_1D8 = pSaveState->field_58;
+    pGlukkon->field_C0_path_number = pSaveState->field_18_path;
+    pGlukkon->field_C2_lvl_number = pSaveState->field_1A_level;
+    pGlukkon->field_CC_sprite_scale = pSaveState->field_1C_sprite_scale;
+
+    pGlukkon->field_D0_r = pSaveState->field_20_r;
+    pGlukkon->field_D2_g = pSaveState->field_22_g;
+    pGlukkon->field_D4_b = pSaveState->field_24_b;
+
+    pGlukkon->field_1A0 = pSaveState->field_20_r; //todo delete if nothing uses it
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.cpp]: isn't it used in vRender?

File: Source/AliveLib/Glukkon.cpp [hunk 4]
```diff
@@ -119,6 +129,130 @@ TintEntry stru_5546B4[18] =
...
+    pGlukkon->field_10C_health = pSaveState->field_30_health;
+    pGlukkon->field_106_current_motion = pSaveState->field_34_current_motion;
+    pGlukkon->field_108_next_motion = pSaveState->field_36_next_motion;
+    pGlukkon->field_F8_LastLineYPos = FP_FromInteger(pSaveState->field_38);
+    pGlukkon->field_114_flags.Set(Flags_114::e114_Bit9);
+    pGlukkon->field_1D4_timer = pSaveState->field_54;
+    pGlukkon->field_104_collision_line_type = pSaveState->field_3A_line_type;
+    pGlukkon->field_214_tlv_info = pSaveState->field_44_tlvInfo;
+    pGlukkon->SetBrain(sGlukkon_ai_table_5544A0[pSaveState->field_48_brain_state_idx]);
+    pGlukkon->field_210 = (__int16) pSaveState->field_50;
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.cpp]: redundant cast

File: Source/AliveLib/Glukkon.cpp [hunk 5]
```diff
@@ -119,6 +129,130 @@ TintEntry stru_5546B4[18] =
...
+    pGlukkon->field_1E2_bUnknown = pSaveState->field_5E;
+    pGlukkon->field_1E4_level = pSaveState->field_60_level;
+    pGlukkon->field_1E6_path = pSaveState->field_62_path;
+    pGlukkon->field_1E8_camera = pSaveState->field_64_camera;
+    pGlukkon->field_1EA_speak = pSaveState->field_66_speak;
+    pGlukkon->field_1E0 = ( __int16 ) pSaveState->field_68;
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.cpp]: ditto

File: Source/AliveLib/Glukkon.cpp [hunk 6]
```diff
@@ -119,6 +129,130 @@ TintEntry stru_5546B4[18] =
...
+    pGlukkon->field_1DC = pSaveState->field_6C;
+    pGlukkon->field_1F0 = pSaveState->field_70;
+    pGlukkon->field_1F4 = pSaveState->field_74;
+    pGlukkon->field_1F8 = pSaveState->field_78;
+    pGlukkon->field_1FC = ( __int16 ) pSaveState->field_7C;
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.cpp]: ditto

File: Source/AliveLib/Glukkon.cpp [hunk 7]
```diff
@@ -199,6 +332,89 @@ void Glukkon::VScreenChanged()
...
+    pSaveState->field_2A_current_frame = this->field_20_animation.field_92_current_frame;
+    pSaveState->field_2C_frame_change_counter = this->field_20_animation.field_E_frame_change_counter;
+    pSaveState->field_26_flipX = this->field_20_animation.field_4_flags.Get(AnimFlags::eBit5_FlipX);
+    pSaveState->field_2E = this->field_20_animation.field_4_flags.Get(AnimFlags::eBit3_Render);
+    pSaveState->field_2F = this->field_6_flags.Get(BaseGameObject::Options::eDrawable);
+    pSaveState->field_30_health = this->field_10C_health;
+    pSaveState->field_34_current_motion = this->field_106_current_motion;
+    pSaveState->field_36_next_motion = this->field_108_next_motion;
+    pSaveState->field_38 = FP_GetExponent(this->field_F8_LastLineYPos);
+    auto collisionLine = this->field_100_pCollisionLine;
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.cpp]: pointless local var

File: Source/AliveLib/Glukkon.cpp [hunk 8]
```diff
@@ -199,6 +332,89 @@ void Glukkon::VScreenChanged()
...
+    if (collisionLine)
+    {
+        pSaveState->field_3A_line_type = collisionLine->field_8_type;
+    }
+    else
+    {
+        pSaveState->field_3A_line_type = -1;
+    }
+    pSaveState->field_40_bIsActiveChar = this == static_cast<Glukkon*>(sControlledCharacter_5C1B8C);
+    pSaveState->field_44_tlvInfo = this->field_214_tlv_info;
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.cpp]: redundant this->

File: Source/AliveLib/Glukkon.cpp [hunk 9]
```diff
@@ -199,6 +332,89 @@ void Glukkon::VScreenChanged()
...
+    {
+        if (BrainIs(fn))
+        {
+            pSaveState->field_48_brain_state_idx = idx;
+            break;
+        }
+        idx++;
+    }
+
+    pSaveState->field_50 = this->field_210;
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.cpp]: ditto for all these

File: Source/AliveLib/Glukkon.hpp [hunk 10]
```diff
@@ -92,15 +92,75 @@ class Glukkon;
...
+ALIVE_ASSERT_SIZEOF_ALWAYS(Glukkon_SaveState, 144);
+
 class Glukkon : public BaseAliveGameObject
 {
 public:
     EXPORT Glukkon* ctor_43F030(Path_Glukkon* pTlv, int tlvInfo);
 
     virtual BaseGameObject* VDestructor(signed int flags) override;
 
+    EXPORT static signed int CC Glukkon::CreateFromSaveState_442830(const BYTE* a1);
```
Comment (paulsapps) [inline Source/AliveLib/Glukkon.hpp]: the others are int not explicit signed int - param a1 should be renamed to pBuffer

File: Source/AliveLib/Glukkon.cpp (hunk 3 above)
Comment (Varrok) [reply Source/AliveLib/Glukkon.cpp]: Possibly, but vRender isn't done yet. I'll handle it, anyway.

**Pull request merged**